django/contrib/admin/paginations.py

```diff
@@ -0,0 +1,116 @@
+from django.contrib.admin.options import IncorrectLookupParameters
+from django.core.paginator import InvalidPage
+from django.template.defaulttags import querystring
+from django.utils.html import format_html, format_html_join
+from django.utils.safestring import mark_safe
+
+ALL_VAR = "all"
+PAGE_VAR = "p"
+
+
+class Pagination:
+    def __init__(
+        self,
+        request,
+        model,
+        list_per_page,
+        list_max_show_all,
+        queryset,
+        model_admin,
+    ):
+        self.model = model
+        self.opts = model._meta
+        self.list_per_page = list_per_page
+        self.list_max_show_all = list_max_show_all
+        self.model_admin = model_admin
+        try:
+            self.page_num = int(request.GET.get(PAGE_VAR, 1))
+        except ValueError:
+            self.page_num = 1
+        self.show_all = ALL_VAR in request.GET
+        self.filter_params = request.GET.copy()
+        self.queryset = queryset
+        self.setup(request)
+
+    @property
+    def pagination_required(self):
+        return (not self.show_all or not self.can_show_all) and self.multi_page
+
+    @property
+    def page_range(self):
+        return (
+            self.paginator.get_elided_page_range(self.page_num)
+            if self.pagination_required
+            else []
+        )
+
+    @property
+    def render_all_pages(self):
+        """
+        Generate the HTML for all page links in the pagination.
+        """
```

Comment:
This name doesn't sound property-like. It sounds more like it does something, rather than representing something. Maybe something like:
```suggestion
    def all_rendered_pages(self):
        """HTML for all the pages in the pagination."""

Reply:
Since it is provided as a property rather than a method, the name you suggested is more appropriate.
Hmm, but I also think it might be better if it were provided as a method rather than a property 🧐